[distributed] NCCLflight recorder timeout fix

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114647
* __->__ #114804

Because isCompleted() returns true on an exception, a timeout exception
will cause the flight recorder to consider the event completed even though it timed out.

This changes the logic to explicitly query the completion events on "retirement"
when the work item leaves the workMetaList. We mark events as retired so
we can distinguish between an event still in the queue but not completed and one
that timed out.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

zdevito committed:
[distributed] NCCLflight recorder timeout fix

Because isCompleted() returns true on an exception, a timeout exception
will cause the flight recorder to consider the event completed even though it timed out.

This changes the logic to explicitly query the completion events on "retirement"
when the work item leaves the workMetaList. We mark events as retired so
we can distinguish between an event still in the queue but not completed and one
that timed out.

[ghstack-poisoned]

zdevito committed:
Update on "[distributed] NCCLflight recorder timeout fix"


Because isCompleted() returns true on an exception, a timeout exception
will cause the flight recorder to consider the event completed even though it timed out.

This changes the logic to explicitly query the completion events on "retirement"
when the work item leaves the workMetaList. We mark events as retired so
we can distinguish between an event still in the queue but not completed and one
that timed out.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -510,6 +517,8 @@ struct NCCLTraceBuffer {
       dict.insert(input_sizes_s, read_sizes(e.input_dims_));
       dict.insert(output_sizes_s, read_sizes(e.output_dims_));
       dict.insert(state_s, e.state_);
+      dict.insert(retired_s, e.retired_);
```
is adding 'retired' to the log really needed?  ideally we'd just avoid putting 'completed' in the log as the state for these cases, and then not need this.  or what's the full set of combinations that can happen?

zdevito replied:
I think when everything is working inside the nccl bindings it doesn't matter too much, but knowing which entries are still in the workMetaList_ and were queried, vs which were already removed might be a useful signal on its own for understanding the state when a timeout happened.

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -406,6 +408,36 @@ struct NCCLTraceBuffer {
   }
 
+  void retire(Entry& r) {
+    r.retired_ = true;
```
can you document the assumptions for this function? I think its important that the work obj is still alive when called, which is guaranteed by the call-site inside the workCleanupLoop right before deletion can happen.  (but i think it's not guaranteed by the other call-site inside dump())

zdevito replied:
I split it in `update_state` which does only what it is named. And then put the stuff that changes other fields directly into the callers.

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -349,6 +349,8 @@ struct NCCLTraceBuffer {
     c10::SmallVector<int, 4> input_dims_;
     c10::SmallVector<int, 4> output_dims_;
     c10::SmallVector<int64_t, 8> sizes_; // flattened from inputs, outputs
+    bool retired_ = false; // is this work entry no longer in the workMetaList_?
```
technically calling 'retire_id()' from inside the dump function would always mark work's as retired, even if they are still inside the workMetaList, wouldn't it?

zdevito replied:
Yeah, let me fix that.

zdevito committed:
Update on "[distributed] NCCLflight recorder timeout fix"


Because isCompleted() returns true on an exception, a timeout exception
will cause the flight recorder to consider the event completed even though it timed out.

This changes the logic to explicitly query the completion events on "retirement"
when the work item leaves the workMetaList. We mark events as retired so
we can distinguish between an event still in the queue but not completed and one
that timed out.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu tianyu-l

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -414,44 +443,21 @@ struct NCCLTraceBuffer {
-        r.end_ = nullptr;
-      }
+      update_state(r);
+      r.start_ = r.end_ = nullptr;
```
why clear the events here?  (if its not actually retired, and if we wanted to dump again later, this would be counterproductive?)

wconstab reviewed (approved):
lgtm! thanks for fixing.

zdevito commented:
@pytorchbot merge
